packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -1449,10 +1458,37 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         }
     }
 
+    // Track how many times the container tries to replay its pending states.
+    // This happens when the connection state is changed.
+    // We reset the internal counter when we are able to process a local op.
+    // If this counter reaches a max, it's a good indicator that the container
+    // is not making progress and it is stuck in a retry loop.
+    private arePendingStatesDrained(): boolean {
+        if (this.maxConsecutiveReplays < 0) {
+            // Feature disabled, we assume we are always making progress
+            return true;
+        }
+
+        return !this.pendingStateManager.hasPendingMessages()
+            || ++this.consecutiveReplays < this.maxConsecutiveReplays;
+    }
+
+    private resetPendingStateReplayCount() {
+        this.consecutiveReplays = 0;
+    }
+
     private replayPendingStates() {
         // We need to be able to send ops to replay states
         if (!this.canSendOps()) { return; }
 
+        if (!this.arePendingStatesDrained()) {
```

Comment:
I'd move this block out to setConnectionState(). This function is called in multiple cases, including read-only transitions - I think it would be wrong to run it in those case.

Reply:
good point